packages/core/src/network/network-peers.ts

```diff
@@ -1,180 +1,103 @@
     return this.peers.findIndex((entry: Entry) => entry.id.equals(peer)) >= 0
   }
 
-  hasBlacklisted(peer: PeerId): boolean {
-    return this.deletedPeers.findIndex((entry: BlacklistedEntry) => entry.id.equals(peer)) >= 0
-  }
-
-  top(n: number): Entry[] {
-    return heap.heaptop(this.peers, n, this.compare)
-  }
-
-  pop(): Entry {
-    return heap.heappop(this.peers, this.compare)
-  }
-
-  blacklistPeer(peer: PeerId): number {
-    verbose('blacklisting', peer.toB58String())
-    const entry = {
-      id: peer,
-      deletedAt: Date.now()
-    }
-
-    // (Efficiently) pushes peer information into blacklist
-    const THRESHOLD_TIMEOUT = Date.now() - BLACKLIST_TIMEOUT
-    const blacklistIndex = this.deletedPeers.findIndex((e: BlacklistedEntry) => e.id.equals(peer))
-    if (blacklistIndex >= 0) {
-      this.deletedPeers[blacklistIndex] = entry
-      heap.heapify(this.deletedPeers, this.compareBlackList)
-    } else if (this.deletedPeers.length == 0) {
-      this.deletedPeers.push(entry)
-    } else if (this.deletedPeers.length > 0) {
-      if (heap.heaptop(this.deletedPeers, 1)[0].deletedAt > THRESHOLD_TIMEOUT) {
-        heap.heapreplace(this.deletedPeers, entry, this.compareBlackList)
-      } else {
-        heap.heappush(this.deletedPeers, entry, this.compareBlackList)
-      }
-    }
-
-    // Keeps the blacklist up-to-date
-    this.cleanupBlacklist(THRESHOLD_TIMEOUT)
-
-    // Removes the peer information from our peerstore
-    const index = this.peers.findIndex((e: Entry) => e.id.equals(entry.id))
-    if (index >= 0) {
-      if (index == this.peers.length - 1) {
-        this.peers.pop()
-      } else {
-        this.peers[index] = this.peers.pop()
-      }
-      heap.heapify(this.peers, this.compare)
-    }
-
-    return this.deletedPeers.length
+  public debugLog() {
+    log(`current nodes:`)
+    this.peers.forEach((e: Entry) => log(`id: ${e.id.toB58String()}, q: ${this.qualityOf(e.id)}`))
   }
 
-  wipeBlacklist(): void {
-    verbose('wiping blacklist')
-    this.deletedPeers = []
+  public containsOlderThan(ts): boolean {
+    return this.peers.length > 0 && heap.heaptop(this.peers, 1, this.compareLastPing)[0].lastSeen < ts
   }
 
-  get length(): number {
-    return this.peers.length
+  public lastSeen(peer: PeerId): number {
+    return this.find(peer).lastSeen
   }
 
-  cleanupBlacklist(THRESHOLD_TIMEOUT: number = Date.now() - BLACKLIST_TIMEOUT) {
-    while (
-      this.deletedPeers.length > 0 &&
-      heap.heaptop(this.deletedPeers, 1, this.compareBlackList)[0].deletedAt < THRESHOLD_TIMEOUT
-    ) {
-      heap.heappop(this.deletedPeers, this.compareBlackList)
+  // @returns a float between 0 (completely unreliable) and 1 (completely
+  // reliable) estimating the quality of service of a peer's network connection
+  public qualityOf(peer: PeerId): number {
+    let entry = this.find(peer)
+    if (entry && entry.heartbeatsSent > 0) {
+      return entry.heartbeatsSuccess / entry.heartbeatsSent
```

Comment:
@SCBuergel Here we have a point of entry for evaluating the quality of a node based on their heartbeats. Pretty much what we initially discussed about a ping bot but now is baked in core.